File: samples/RazorPagesMovie/Models/Movie.cs
using DevZest.Data;
using DevZest.Data.Annotations;
using DevZest.Data.SqlServer;

namespace RazorPagesMovie.Models
{
    public class Movie : Model<Movie.PK>
    {
        public sealed class PK : CandidateKey
        {
            public PK(_Int32 id) : base(id)
            {
            }
        }

        protected sealed override PK CreatePrimaryKey()
        {
            return new PK(ID);
        }

        public class Key : Key<PK>
        {
            static Key()
            {
                Register((Key _) => _.ID, _ID);
            }

            protected sealed override PK CreatePrimaryKey()
            {
                return new PK(ID);
            }

            public _Int32 ID { get; private set; }
        }

        public static readonly Mounter<_Int32> _ID = RegisterColumn((Movie _) => _.ID);
        public static readonly Mounter<_String> _Title = RegisterColumn((Movie _) => _.Title);
        public static readonly Mounter<_DateTime> _ReleaseDate = RegisterColumn((Movie _) => _.ReleaseDate);
        public static readonly Mounter<_String> _Genre = RegisterColumn((Movie _) => _.Genre);
        public static readonly Mounter<_Decimal> _Price = RegisterColumn((Movie _) => _.Price);

        [Identity(1, 1)]
        public _Int32 ID { get; private set; }

        [StringLength(60, MinimumLength = 3)]
        [Required]
        [SqlNVarChar(60)]
        public _String Title { get; private set; }

        [Display(Name = "Release Date")]
        [SqlDate]
        public _DateTime ReleaseDate { get; private set; }

        [RegularExpression(@"^[A-Z]+[a-zA-Z""'\s-]*$")]
        [Required]
        [StringLength(30)]
        [SqlNVarChar(30)]
        public _String Genre { get; private set; }

        [SqlMoney]
        [Required]
        public _Decimal Price { get; private set; }
    }
}

File: samples.vb/Tutorial/Movies/Movie.vb
Imports DevZest.Data
Imports DevZest.Data.Annotations
Imports DevZest.Data.SqlServer

Public Class Movie
    Inherits Model(Of Movie.PK)

    Public NotInheritable Class PK
        Inherits CandidateKey

        Public Sub New(id As _Int32)
            MyBase.New(id)
        End Sub
    End Class

    Protected NotOverridable Overrides Function CreatePrimaryKey() As PK
        Return New PK(ID)
    End Function

    Public Class Key
        Inherits Key(Of PK)

        Shared Sub New()
            Register(Function(x As Key) x.ID, _ID)
        End Sub

        Protected NotOverridable Overrides Function CreatePrimaryKey() As PK
            Return New PK(ID)
        End Function

        Private m_ID As _Int32

        Public Property ID As _Int32
            Get
                Return m_ID
            End Get
            Private Set
                m_ID = Value
            End Set
        End Property
    End Class

    Public Shared ReadOnly _ID As Mounter(Of _Int32) = RegisterColumn(Function(x As Movie) x.ID)
    Public Shared ReadOnly _Title As Mounter(Of _String) = RegisterColumn(Function(x As Movie) x.Title)
    Public Shared ReadOnly _ReleaseDate As Mounter(Of _DateTime) = RegisterColumn(Function(x As Movie) x.ReleaseDate)
    Public Shared ReadOnly _Genre As Mounter(Of _String) = RegisterColumn(Function(x As Movie) x.Genre)
    Public Shared ReadOnly _Price As Mounter(Of _Decimal) = RegisterColumn(Function(x As Movie) x.Price)

    Private m_ID As _Int32
    <Identity>
    Public Property ID As _Int32
        Get
            Return m_ID
        End Get
        Private Set
            m_ID = Value
        End Set
    End Property

    Private m_Title As _String
    <StringLength(60, MinimumLength:=3)>
    <Required>
    <SqlNVarChar(60)>
    Public Property Title As _String
        Get
            Return m_Title
        End Get
        Private Set
            m_Title = Value
        End Set
    End Property

    Private m_ReleaseDate As _DateTime
    <Display(Name:="Release Date")>
    <SqlDate>
    <Required>
    Public Property ReleaseDate As _DateTime
        Get
            Return m_ReleaseDate
        End Get
        Private Set
            m_ReleaseDate = Value
        End Set
    End Property

    Private m_Genre As _String
    <RegularExpression("^[A-Z]+[a-zA-Z""'\s-]*$")>
    <Required>
    <StringLength(30)>
    <SqlNVarChar(30)>
    Public Property Genre As _String
        Get
            Return m_Genre
        End Get
        Private Set
            m_Genre = Value
        End Set
    End Property

    Private m_Price As _Decimal
    <SqlMoney>
    <Required>
    Public Property Price As _Decimal
        Get
            Return m_Price
        End Get
        Private Set
            m_Price = Value
        End Set
    End Property
End Class
